### ISSUE TITLE
sourceMappingURL detection can fail in some cases

### ISSUE THREAD

**Issue by jampekka** [systemjs/builder#606]
`lib/builder.js` overrides the fetch-hook to look for source maps. However, in some (albeit rare) cases, the regex matches eg. code that tries to print source maps, which leads the fetcher to try to read from a non-existant file, causing the build to fail. For example, the builder can't bundle the file `lib/output.js` because of this.

EDIT: This is wrong. See below for a proper reproduction.

Can be reproduced by the following:
```javascript
var Builder = require('systemjs-builder');
builder = new Builder('./');
builder.build('./node_modules/systemjs-builder/lib/output.js');
```

A quick workaround would be to catch the error from reading the source map. However, the detection could also be more robust, as at least the sourcemaps spec is quite clear that sourceMappingUrl should be at the end of the file:
```
The generated code may include a line at the end of the source, with the following form:

//# sourceMappingURL=<url>
```
https://docs.google.com/document/d/1U1RGAehQwRypUTovF1KRlpiOFze0b-_2gc6fAH0KY0k/edit#

https://github.com/systemjs/builder/blob/271934fe5473bdccf975fb49da8574fb350b7c57/lib/builder.js#L199

**Comment by asapach** [comment]
Are you testing it with `0.15.18`? Can you try `0.15.17`?

**Comment by jampekka** [comment]
Same thing with `0.15.17`:

```console
$ npm install systemjs-builder@0.15.17
...
$ node build.js
Unhandled rejection Error on fetch for path at file:///home/jampekka/tmp/systemjsbug/path
	Loading node_modules/systemjs-builder/lib/output.js
	Error: ENOENT: no such file or directory, open '/home/jampekka/tmp/systemjsbug/path'
    at Error (native)
From previous event:
    at SystemJSNodeLoader.<anonymous> (/home/jampekka/tmp/systemjsbug/node_modules/systemjs-builder/node_modules/systemjs/dist/system.src.js:1788:12)
    at SystemJSNodeLoader.<anonymous> (/home/jampekka/tmp/systemjsbug/node_modules/systemjs-builder/node_modules/systemjs/dist/system.src.js:2811:22)
    at SystemJSNodeLoader.<anonymous> (/home/jampekka/tmp/systemjsbug/node_modules/systemjs-builder/node_modules/systemjs/dist/system.src.js:3385:20)
    at SystemJSNodeLoader.<anonymous> (/home/jampekka/tmp/systemjsbug/node_modules/systemjs-builder/node_modules/systemjs/dist/system.src.js:3661:18)
    at SystemJSNodeLoader.<anonymous> (/home/jampekka/tmp/systemjsbug/node_modules/systemjs-builder/node_modules/systemjs/dist/system.src.js:4050:18)
    at SystemJSNodeLoader.<anonymous> (/home/jampekka/tmp/systemjsbug/node_modules/systemjs-builder/node_modules/systemjs/dist/system.src.js:4513:22)
    at SystemJSNodeLoader.fetch (/home/jampekka/tmp/systemjsbug/node_modules/systemjs-builder/node_modules/systemjs/dist/system.src.js:4781:20)
    at SystemJSNodeLoader.loader.fetch (/home/jampekka/tmp/systemjsbug/node_modules/systemjs-builder/lib/builder.js:178:59)
    at /home/jampekka/tmp/systemjsbug/node_modules/systemjs-builder/lib/trace.js:425:23
    at processImmediate [as _immediateCallback] (timers.js:383:17)
```

With `build.js` being:

```javascript
global.Promise = require('bluebird');
Promise.longStackTraces();
var Builder = require('systemjs-builder');
builder = new Builder('./');
builder.bundle('./node_modules/systemjs-builder/lib/output.js');
```

**Comment by jampekka** [comment]
Sorry, the test case fails before this bug on an unresolved `require`. I'll update the case.

**Comment by jampekka** [comment]
@asapach Actually the bug doesn't occur with `0.15.17`, sorry for my hasty claim. A fixed test case:
build.js:
```javascript
global.Promise = require('bluebird');
Promise.longStackTraces();
var Builder = require('systemjs-builder');
builder = new Builder('./');
builder.bundle('./test.js');
```

test.js:
```javascript
function addSourceMap(source, sourceMapFileName) {
	return source += '\n//# sourceMappingURL=' + sourceMapFileName;
}
```

Fails on `0.15.18` with:

```console
$ nodejs build.js
Unhandled rejection Error: Error on fetch for test.js at file:///home/jampekka/tmp/systemjsbug/test.js
	EISDIR: illegal operation on a directory, read
    at Error (native)
    at /home/jampekka/tmp/systemjsbug/node_modules/systemjs-builder/lib/builder.js:210:32
    at SystemJSNodeLoader.loader.fetch (/home/jampekka/tmp/systemjsbug/node_modules/systemjs-builder/lib/builder.js:196:6)
    at /home/jampekka/tmp/systemjsbug/node_modules/systemjs-builder/lib/trace.js:425:23
    at /home/jampekka/tmp/systemjsbug/node_modules/systemjs-builder/lib/trace.js:418:8
    at /home/jampekka/tmp/systemjsbug/node_modules/systemjs-builder/lib/trace.js:388:6
    at processImmediate [as _immediateCallback] (timers.js:383:17)
    at Trace.getLoadRecord (/home/jampekka/tmp/systemjsbug/node_modules/systemjs-builder/lib/trace.js:131:4)
    at Trace.getAllLoadRecords (/home/jampekka/tmp/systemjsbug/node_modules/systemjs-builder/lib/trace.js:552:15)
    at /home/jampekka/tmp/systemjsbug/node_modules/systemjs-builder/lib/trace.js:54:19
    at Trace.traceCanonical (/home/jampekka/tmp/systemjsbug/node_modules/systemjs-builder/lib/trace.js:52:4)
    at /home/jampekka/tmp/systemjsbug/node_modules/systemjs-builder/lib/arithmetic.js:243:35
    at /home/jampekka/tmp/systemjsbug/node_modules/systemjs-builder/lib/arithmetic.js:232:20
    at Array.reduce (native)
    at processExpandedOperations (/home/jampekka/tmp/systemjsbug/node_modules/systemjs-builder/lib/arithmetic.js:231:35)
    at exports.traceExpression (/home/jampekka/tmp/systemjsbug/node_modules/systemjs-builder/lib/arithmetic.js:229:8)
    at /home/jampekka/tmp/systemjsbug/node_modules/systemjs-builder/lib/builder.js:639:14
    at /home/jampekka/tmp/systemjsbug/node_modules/systemjs-builder/lib/builder.js:638:6
    at processImmediate [as _immediateCallback] (timers.js:383:17)
    at Builder.bundle (/home/jampekka/tmp/systemjsbug/node_modules/systemjs-builder/lib/builder.js:630:4)
    at Object.<anonymous> (/home/jampekka/tmp/systemjsbug/build.js:5:9)
    at Module._compile (module.js:409:26)
    at Object.Module._extensions..js (module.js:416:10)
    at Module.load (module.js:343:32)
    at Function.Module._load (module.js:300:12)
    at Function.Module.runMain (module.js:441:10)
    at startup (node.js:139:18)
    at node.js:968:3
From previous event:
    at /home/jampekka/tmp/systemjsbug/node_modules/systemjs-builder/lib/trace.js:483:13
From previous event:
    at /home/jampekka/tmp/systemjsbug/node_modules/systemjs-builder/lib/trace.js:388:6
    at processImmediate [as _immediateCallback] (timers.js:383:17)
From previous event:
    at Trace.getLoadRecord (/home/jampekka/tmp/systemjsbug/node_modules/systemjs-builder/lib/trace.js:131:4)
    at Trace.getAllLoadRecords (/home/jampekka/tmp/systemjsbug/node_modules/systemjs-builder/lib/trace.js:552:15)
    at /home/jampekka/tmp/systemjsbug/node_modules/systemjs-builder/lib/trace.js:54:19
From previous event:
    at Trace.traceCanonical (/home/jampekka/tmp/systemjsbug/node_modules/systemjs-builder/lib/trace.js:52:4)
    at /home/jampekka/tmp/systemjsbug/node_modules/systemjs-builder/lib/arithmetic.js:243:35
From previous event:
    at /home/jampekka/tmp/systemjsbug/node_modules/systemjs-builder/lib/arithmetic.js:232:20
    at Array.reduce (native)
    at processExpandedOperations (/home/jampekka/tmp/systemjsbug/node_modules/systemjs-builder/lib/arithmetic.js:231:35)
From previous event:
    at exports.traceExpression (/home/jampekka/tmp/systemjsbug/node_modules/systemjs-builder/lib/arithmetic.js:229:8)
    at /home/jampekka/tmp/systemjsbug/node_modules/systemjs-builder/lib/builder.js:639:14
From previous event:
    at /home/jampekka/tmp/systemjsbug/node_modules/systemjs-builder/lib/builder.js:638:6
    at processImmediate [as _immediateCallback] (timers.js:383:17)
From previous event:
    at Builder.bundle (/home/jampekka/tmp/systemjsbug/node_modules/systemjs-builder/lib/builder.js:630:4)
    at Object.<anonymous> (/home/jampekka/tmp/systemjsbug/build.js:5:9)
    at Module._compile (module.js:409:26)
    at Object.Module._extensions..js (module.js:416:10)
    at Module.load (module.js:343:32)
    at Function.Module._load (module.js:300:12)
    at Function.Module.runMain (module.js:441:10)
    at startup (node.js:139:18)
    at node.js:968:3
```

**Comment by asapach** [comment]
@guybedford, maybe you shouldn't have gotten rid of `source-map-url`? :)

**Comment by guybedford** [comment]
Fixed in https://github.com/systemjs/builder/commit/3d95d8dd9ecbf4dfbb97766bfd9a18ff37232132. @asapach I had to alter the regex because it was catching `/* sourceMappingURL=x */` cases which was breaking on a comment in Babel's build.

**Comment by guybedford** [comment]
Released in 0.15.19.
